File: src/Hush.VSCode/src/state/UserSlotMap.ts
import * as vscode from 'vscode';
import { UserSlotCount } from '../colors/builtInColors';

const STORAGE_KEY = 'hush.userSlotMap';

/**
 * Persistent first-seen-category → slot mapping, so a user-defined category
 * keeps the same hotkey across sessions. Mirrors `Options/UserSlotMap.cs` from
 * the VS extension.
 */
export class UserSlotMap {
  private categoryBySlot = new Map<number, string>();
  private slotByCategory = new Map<string, number>();

  constructor(private readonly state: vscode.Memento) {
    const persisted = state.get<Record<string, number>>(STORAGE_KEY, {});
    for (const [cat, slot] of Object.entries(persisted)) {
      if (typeof slot === 'number' && slot >= 1 && slot <= UserSlotCount) {
        this.slotByCategory.set(cat.toLowerCase(), slot);
        this.categoryBySlot.set(slot, cat);
      }
    }
  }

  /** Returns the slot for a category, allocating the lowest free slot if it's new. */
  async assign(categoryKey: string): Promise<number | undefined> {
    const k = categoryKey.toLowerCase();
    const existing = this.slotByCategory.get(k);
    if (existing) return existing;
    for (let slot = 1; slot <= UserSlotCount; slot++) {
      if (!this.categoryBySlot.has(slot)) {
        this.slotByCategory.set(k, slot);
        this.categoryBySlot.set(slot, categoryKey);
        await this.persist();
        return slot;
      }
    }
    return undefined; // all slots taken
  }

  categoryForSlot(slot: number): string | undefined {
    return this.categoryBySlot.get(slot);
  }

  private async persist(): Promise<void> {
    const map: Record<string, number> = {};
    for (const [cat, slot] of this.slotByCategory) map[cat] = slot;
    await this.state.update(STORAGE_KEY, map);
  }
}

File: src/Hush.VS/Options/UserSlotMap.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Hush.VS.Options;

/// <summary>
/// Stable mapping between user-defined category keys and the fixed pool of
/// "hush.user1..N" classification types. Persisted as a compact "key1:slot1;key2:slot2"
/// string so the same category keeps the same slot (and therefore the same F&amp;C entry)
/// across sessions.
/// </summary>
internal sealed class UserSlotMap
{
    private readonly Dictionary<string, int> _map = new(StringComparer.OrdinalIgnoreCase);
    private readonly int _slotCount;

    public UserSlotMap(int slotCount = Constants.UserSlotCount)
    {
        _slotCount = slotCount;
    }

    public int? Get(string categoryKey) =>
        _map.TryGetValue(categoryKey, out var s) ? s : (int?)null;

    public int Assign(string categoryKey)
    {
        if (_map.TryGetValue(categoryKey, out var existing)) return existing;
        for (int i = 1; i <= _slotCount; i++)
        {
            if (!_map.Values.Contains(i))
            {
                _map[categoryKey] = i;
                return i;
            }
        }
        // No slots free: alias to slot 1 so the user still sees *something* muted
        // rather than silently dropping the category.
        _map[categoryKey] = 1;
        return 1;
    }

    public string? CategoryForSlot(int slot)
    {
        foreach (var kv in _map)
            if (kv.Value == slot) return kv.Key;
        return null;
    }

    public string Serialize() =>
        string.Join(";", _map.Select(kv => $"{kv.Key}:{kv.Value}"));

    public static UserSlotMap Deserialize(string? raw, int slotCount = Constants.UserSlotCount)
    {
        var m = new UserSlotMap(slotCount);
        if (string.IsNullOrWhiteSpace(raw)) return m;
        foreach (var part in raw!.Split(';'))
        {
            var bits = part.Split(':');
            if (bits.Length == 2 && int.TryParse(bits[1], out var slot))
            {
                m._map[bits[0]] = slot;
            }
        }
        return m;
    }
}
